api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/DrugQueryBuilder.java

```diff
@@ -25,45 +27,63 @@
-      "    join (select CONCAT( '%.', CAST(id as STRING), '%') as path from `${projectId}.${dataSetId}.criteria` where concept_id in unnest(${parentConceptIds})) b \n" +
-      "    on a.path like b.path\n" +
-      "    and is_group = 0\n" +
-      "    and is_selectable = 1\n" +
-      "    and type = 'DRUG'\n" +
-      "    and subtype = 'ATC')\n";
+  private static final String CHILD_ONLY_TEMPLATE =
+    "concept_id in unnest(${childConceptIds})\n";
 
-  private static final String UNION_TEMPLATE = " union all\n";
+  private static final String PARENT_CRITERIA = "select a.concept_id from\n" +
+    "`${projectId}.${dataSetId}.criteria` a\n" +
+    "join (select CONCAT( '%.', CAST(id as STRING), '%') as path\n" +
+    "from `${projectId}.${dataSetId}.criteria`\n" +
+    "where concept_id in unnest(${parentConceptIds})) b\n" +
+    "on a.path like b.path\n" +
+    "and is_group = 0\n" +
+    "and is_selectable = 1\n" +
+    "and type = 'DRUG'\n" +
+    "and subtype = 'ATC'";
 
-  private static final String ENCOUNTERS_SQL_TEMPLATE = "${encounterSql}";
+  private static final String BOTH_TEMPLATE =
+    "concept_id in (" + PARENT_CRITERIA + " or\n" +
+      CHILD_ONLY_TEMPLATE + ")\n";
+
+  private static final String PARENT_ONLY_TEMPLATE =
+    "concept_id in (" + PARENT_CRITERIA + ")\n";
 
   @Override
-  public String buildQuery(Map<String, QueryParameterValue> queryParams, QueryParameters parameters) {
-    from(parametersEmpty()).test(parameters.getParameters()).throwException(EMPTY_MESSAGE, PARAMETERS);
-    ListMultimap<String, Long> paramMap = getMappedParameters(parameters.getParameters());
-    List<String> queryParts = new ArrayList<>();
-    for (String key : paramMap.keySet()) {
-      Long[] conceptIds = paramMap.get(key).stream().toArray(Long[]::new);
-      String namedParameter = addQueryParameterValue(queryParams, QueryParameterValue.array(conceptIds, Long.class));
-      if ("Children".equals(key)) {
-        queryParts.add(DRUG_CHILD_SQL_TEMPLATE.replace("${childConceptIds}", "@" + namedParameter));
-      } else {
-        queryParts.add(DRUG_PARENT_SQL_TEMPLATE.replace("${parentConceptIds}", "@" + namedParameter));
-      }
-    }
+  public String buildQuery(Map<String, QueryParameterValue> queryParams,
+                           SearchGroupItem searchGroupItem,
+                           String mention) {
+    from(parametersEmpty()).test(searchGroupItem.getSearchParameters()).throwException(EMPTY_MESSAGE, PARAMETERS);
+    ListMultimap<String, Long> paramMap = getMappedParameters(searchGroupItem.getSearchParameters());
+    StringBuilder baseSql = new StringBuilder(DRUG_SQL_TEMPLATE);
+    StringBuilder conceptIdSql = new StringBuilder();
 
-    String drugSql = String.join(UNION_TEMPLATE, queryParts) + ENCOUNTERS_SQL_TEMPLATE;
-    return buildModifierSql(drugSql, queryParams, parameters.getModifiers());
+    Long[] parentIds = paramMap.get(PARENT).stream().toArray(Long[]::new);
```

Comment:
Could you add a comment explaining why these cases need to be split out like this?